src/platform/backends/shared/base_snapshot.cpp

```diff
@@ -178,11 +179,18 @@ mp::BaseSnapshot::BaseSnapshot(const QString& filename, VirtualMachine& vm, cons
 
 mp::BaseSnapshot::BaseSnapshot(const QJsonObject& json, VirtualMachine& vm, const VirtualMachineDescription& desc)
     : BaseSnapshot{
-          json["name"].toString().toStdString(),                   // name
-          json["comment"].toString().toStdString(),                // comment
-          json["cloud_init_instance_id"].toString().toStdString(), // instance id from cloud init
-          find_parent(json, vm),                                   // parent
-          json["index"].toInt(),                                   // index
+          json["name"].toString().toStdString(),    // name
+          json["comment"].toString().toStdString(), // comment
+          // When it does not contain cloud_init_instance_id, it signifies that the legacy snapshot does not have the
+          // item and it needs to fill cloud_init_instance_id with the current value. The current value equals to the
+          // value at snapshot time because cloud_init_instance_id has been an immutable variable up to this point.
+          json.contains("cloud_init_instance_id")
+              ? json["cloud_init_instance_id"].toString().toStdString()
+              : MP_CLOUD_INIT_FILE_OPS.get_instance_id_from_cloud_init(
+                    std::filesystem::path{vm.instance_directory().absolutePath().toStdString()} /
+                    "cloud-init-config.iso"), // instance id from cloud init
```

Comment:
Hey @georgeliao, I believe it would help readability if this was extracted into a helper function. Would you mind doing that?

Reply:
That can be done.